Virtual thermostat

DB-CL opened the pull request:
## Description:

The `virtual_thermostat` climate platform is a new thermostat implementation in Home Assistant. It handles things differently from its ancestor the `generic_thermostat` being more compliant with the new climate 1.0 and supposed to be closer to a real thermostat. For example, in one component, you can have both AC and heating, avoiding to create multiple instances to control the climate in one area. It also supports an _Away mode_, an _Eco mode_ and a _Comfort mode_

I started a new implementation to not break the `generic_thermostat` implementation, but I think it could be a good replacement in the future, since the code of the `generic_thermostat` is not well maintained and not very compliant with the new climate 1.0. 

It should solve, or at least propose an alternative solution for these issues :
https://github.com/home-assistant/home-assistant/issues/27350
https://github.com/home-assistant/home-assistant/issues/27069
https://github.com/home-assistant/home-assistant/issues/21125
https://github.com/home-assistant/home-assistant/issues/20154

Please, tell me if you see something wrong with this PR. 

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) :** home-assistant/home-assistant.io#10831
**Pull request with frontend modifications :** home-assistant/home-assistant-polymer/pull/3718

## Example entry for `configuration.yaml` (if applicable):
```yaml
default_config:

frontend:

logger:
  default: info
  logs:
    homeassistant.components.virtual_thermostat: debug

input_boolean:
  fake_heater_input:
    name: Fake heater input
    initial: off
  fake_ac_input:
    name: Fake ac input
    initial: off

input_number:
  fake_sensor_temperature_1:
    initial: 20
    min: -20
    max: 35
    step: 0.5

sensor:
  - platform: template
    sensors:
      fake_sensor_1:
        unit_of_measurement: "degrees"
        value_template: "{{ states('input_number.fake_sensor_temperature_1') | float }}"
    
switch:
  - platform: template
    switches:
      fake_heater_switch:
        value_template: "{{ is_state('input_boolean.fake_heater_input', 'on') }}"
        turn_on:
          service: input_boolean.turn_on
          entity_id: input_boolean.fake_heater_input
        turn_off:
          service: input_boolean.turn_off
          entity_id: input_boolean.fake_heater_input
      fake_ac_switch:
        value_template: "{{ is_state('input_boolean.fake_ac_input', 'on') }}"
        turn_on:
          service: input_boolean.turn_on
          entity_id: input_boolean.fake_ac_input
        turn_off:
          service: input_boolean.turn_off
          entity_id: input_boolean.fake_ac_input

climate:
  - platform: virtual_thermostat
    name: Fake thermostat 4
    sensor: sensor.fake_sensor_1
    initial_hvac_mode: "off"
    initial_preset_mode: "none"
    hysteresis_tolerance_on: 0.5
    hysteresis_tolerance_off: 1
    restore_from_old_state: True
    keep_alive:
      minutes: 3
    min_cycle_duration:
      minutes: 5
    enabled_presets:
      - away
      - eco
      - comfort
    heat:
      entity_id: switch.fake_heater_switch
      min_temp: 15
      max_temp: 24
      initial_target_temp: 19
      away_temp: 12
      eco_shift: -2
      comfort_shift: 2
    cool:
      entity_id: switch.fake_ac_switch
      min_temp: 10
      max_temp: 32
      initial_target_temp: 28
      away_temp: 28
      eco_shift: 4
      comfort_shift: -2
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
_I never managed to get tox working properly with my local installation :( There are a lot of python errors and I think they are not related with my code. I launched pytest, pylint and flake8 manually and everything seems good to go. I hope it's OK_
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [Not concerned] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [Not concerned] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

## Additionnal informations ##

I executed the tests with pytest : `pytest tests/components/virtual_thermostat/test_climate.py --cov=homeassistant/components/virtual_thermostat/ --cov-report term-missing`

And it is reporting a full coverage : 

```
----------- coverage: platform linux, python 3.7.4-final-0 -----------
Name                                                      Stmts   Miss  Cover   Missing
---------------------------------------------------------------------------------------
homeassistant/components/virtual_thermostat/__init__.py       0      0   100%
homeassistant/components/virtual_thermostat/climate.py      369      0   100%
---------------------------------------------------------------------------------------
TOTAL                                                       369      0   100%

```

The linter `pylint homeassistant` is happy with 10.0 and `flake8 homeassistant` reports nothing.

## Breaking Change:

No breaking changes. It's a new component.

DB-CL committed:
WIP virtual thermostat

DB-CL committed:
WIP

DB-CL committed:
100% tests coverage

DB-CL committed:
Manifest, codeowners, typo fix

DB-CL committed:
Lint problem

homeassistant commented:
Hello @DB-CL,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`1d64580a27b89e8c4dd97aebe47427522a10613d`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;27833&#x2F;commits&#x2F;1d64580a27b89e8c4dd97aebe47427522a10613d) - This commit has something that looks like an email address (cyril.dubus@effipilot.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

DB-CL commented:
I looked into the two failures, and I'm afraid I cannot understand what is really going on. Can someone give me some clue ? Thanks

DB-CL committed:
Test file blacked

DB-CL commented:
Ok I missed this https://developers.home-assistant.io/blog/2019/07/31/black.html and for some reason my editor did "blacked" my component, but not my test file.

b4dpxl commented:
@DB-CL I've only had a quick look, I'll have a proper look later. So I apologise, you might have addressed all the below and I've not seen it. I like the "eco" implementation though, I can definitely see a use for that 👍

I see you're skipping triggering the switch when "away", but still setting the temperature. If you're using automations to set the temperature, but have enabled "away", although that won't trigger the heating/cooling immediately, it would potentially be triggered next time the temperature changes?

Example:
away enabled with temperature: 12C
current temperature: 20C - the heating is off
scheduled temperature set via automation to: 19C
current temperature: 20C - heating still off
temperature drops to 18C - heating now turns on

This is why I update and persistent `saved_temperature` in my `virtual_thermostat` if "away".

Also, what happens if the sensor goes offline/unavailable? In the current `generic_thermostat`, the switch will stay in its current state forever, heating or cooling. I've had this at home when a sensor dropped off the network; it got _very_ hot in the house :D I added support to my `virtual_thermostat` to take the component offline if the sensor _or_ switch go unavailable, turn off the switch in that situation (to prevent uncontrolled heating/cooling), and also an optional timeout in case a sensor is still available but stops updating.

Thanks

DB-CL committed:
Add a test case for b4dpxl's question

DB-CL committed:
Update CODEOWNERS

DB-CL commented:
Hi @b4dpxl,

Thank you for having a look into my code !

I rebased and I added a test case to cover your question, can you have a look and tell me if everything looks OK to you ? You can see the patch here : https://github.com/DB-CL/home-assistant/commit/61c65cfe8976b9006bf3963d07f36bca27415856

In short : I think it's behaving the way you want (and the way it should), but if you can read the test and confirm this is what you wanted to see, that would be perfect.

If the sensor goes OFF, the component won't do anything. It's a very good suggestion to add a timeout option to turn OFF everything and I will certainly look into it. I'm wondering if I should wait this PR to be accepted first, to focus on having a code compatible with hass standards, and then start to look into suggestions like yours. Or if I should start working on suggestion right away. I feel like I should wait, what do you think ? I don't know how long it will take for someone with write access to have a look to this PR.

ciotlosm commented:
@DB-CL why create a completely new component instead of contributing to the existing one? Having multiple components that are for similar use cases create more confusion. 

I agree that `generic_thermostat` requires changes, including a new config flow, but it would be best in my opinion if that got an update instead of creating a completely new component.

b4dpxl commented:
> @DB-CL ~why create a completely new component instead of contributing to the existing one?~ Having multiple components that are for similar use cases create more confusion.
> 
> I agree that `generic_thermostat` requires changes, including a new config flow, but it would be best in my opinion if that got an update instead of creating a completely new component. I would be in favor of breaking changes instead of duplicates.
> 
> Update: You actually specified the reason behind a new component, but I still think a debate would be healthy to make sure we don't bring more maintenance. As HA is not yet in 1.0 breaking changes should be the way to go if it removes technical debt for the future.



> Hi @b4dpxl,
> 
> Thank you for having a look into my code !
> 
> I rebased and I added a test case to cover your question, can you have a look and tell me if everything looks OK to you ? You can see the patch here : [DB-CL@61c65cf](https://github.com/DB-CL/home-assistant/commit/61c65cfe8976b9006bf3963d07f36bca27415856)
> 
> In short : I think it's behaving the way you want (and the way it should), but if you can read the test and confirm this is what you wanted to see, that would be perfect.
> 
> If the sensor goes OFF, the component won't do anything. It's a very good suggestion to add a timeout option to turn OFF everything and I will certainly look into it. I'm wondering if I should wait this PR to be accepted first, to focus on having a code compatible with hass standards, and then start to look into suggestions like yours. Or if I should start working on suggestion right away. I feel like I should wait, what do you think ? I don't know how long it will take for someone with write access to have a look to this PR.

Awesome. I'm not in a position to test it at the moment, but I'll try and have a proper look over the weekend :) Thanks

b4dpxl commented:
> @DB-CL ~why create a completely new component instead of contributing to the existing one?~ Having multiple components that are for similar use cases create more confusion.
> 
> I agree that `generic_thermostat` requires changes, including a new config flow, but it would be best in my opinion if that got an update instead of creating a completely new component. I would be in favor of breaking changes instead of duplicates.
> 
> Update: You actually specified the reason behind a new component, but I still think a debate would be healthy to make sure we don't bring more maintenance. As HA is not yet in 1.0 breaking changes should be the way to go if it removes technical debt for the future.

@ciotlosm personally, I started on my own instance of a `virtual_thermostat` after having PR's I created to resolve an issue I raised arbitrarily (it seemed to me) rejected by people who failed to understand the issues it was attempting to resolve, comments like the below from HA contributors, and what appeared to be a general apathy towards the component:

> I have spent too much time in a component I don't have interest in. I will not respond this thread anymore.

`generic_thermostat` was the reason I started using HA in the first place as, despite a few issues which _could_ (but shouldn't have to be) worked around, it gave me exactly what I wanted; an internet-controllable smart thermostat, without the restraints or potential security risks of a third-party managed IoT device.

DB-CL commented:
> @DB-CL ~why create a completely new component instead of contributing to the existing one?~ Having multiple components that are for similar use cases create more confusion.
> 
> I agree that `generic_thermostat` requires changes, including a new config flow, but it would be best in my opinion if that got an update instead of creating a completely new component. I would be in favor of breaking changes instead of duplicates.
> 
> Update: You actually specified the reason behind a new component, but I still think a debate would be healthy to make sure we don't bring more maintenance. As HA is not yet in 1.0 breaking changes should be the way to go if it removes technical debt for the future.

Hi, 

Thank you for looking into my PR ! :+1: 

I agree that we can debate around this specific issue (old vs new component). I chose something (ie: to add a new component instead of replacing the old one) to have something to discuss about. I wouldn't mind at all to totally replace the generic_thermostat.

To start the debate, this is my point of view :

First, the generic thermostat needs a rewrite. You cannot fix it : when you dig into the code, you realize it's not made at all for climate 1.0, and it has a lot of issues, because it has grown in every directions. My component is not 100% different, I pick code from the generic_thermostat. But it's a rewrite anyway : the way it behave internally is not the same as generic_thermostat (I tried to fix it first, and I just waste a couple of hours..)

Second, as b4dpxl stated, there is a general apathy about this component, and I don't think there will be some movement neither some acceptance any time soon. I also think this generic_thermostat is widely used, and the maintainers do not want to change it.

Finally, I personally think it's best to have an old component, then a new one, then to ask ppl to move from the old to the new, then to deprecate the old one. It let people the time to test the new one, make suggestions, explain what is good or bad in the new one, what they like in the old one they cannot do in the new one, etc.. This way, the new one can replace properly the old one. I feel like imposing a lot of changes without letting people to get used to the new component is brutal, and can create a lot of frustration. This third point is a personal point of view and I can understand it is not shared.


I'm not familiar with HA hierarchy. Who can take a decision on this topic ? Because if I need to remake this PR to replace generic_thermostat it's not complicated.

Have a nice day

ciotlosm commented:
> Hi,
> 
> Thank you for looking into my PR ! 👍
> 
> I agree that we can debate around this specific issue (old vs new component). I chose something (ie: to add a new component instead of replacing the old one) to have something to discuss about. I wouldn't mind at all to totally replace the generic_thermostat.
> 

I think that is the best way. There are some cases where you'd go through architecture before committing a lot of hours into writing code. 

> To start the debate, this is my point of view :
> 
> First, the generic thermostat needs a rewrite. You cannot fix it : when you dig into the code, you realize it's not made at all for climate 1.0, and it has a lot of issues, because it has grown in every directions. My component is not 100% different, I pick code from the generic_thermostat. But it's a rewrite anyway : the way it behave internally is not the same as generic_thermostat (I tried to fix it first, and I just waste a couple of hours..)
> 

I agree that `generic_thermostat` needs a rewrite. I also think that it would benefit a lot from having a config flow as well. 

> Second, as b4dpxl stated, there is a general apathy about this component, and I don't think there will be some movement neither some acceptance any time soon. I also think this generic_thermostat is widely used, and the maintainers do not want to change it.
> 

I think there are some potential misconceptions here. The component is complex and the code not that great for current standards (I wrote part of the current code myself). The 'apathy' might be there because it's a big undertake to rewrite the component. 

Yes, the `generic_thermostat` component is widely used but people would be against big updates only if they radically change how the component functions, breaking many of the current setups. I remember the pain when I modified it last time. 

There are two approaches to rewrite: small increments and big bang approaches. The problem with the second is that is really hard for people that already contributed to the component to objectively provide accurate feedback and make sure they there are no big differences in functionality for all the use cases that the previous one provided. 

I tried to re-write tests in a better format, but my first PR changed hundreds of lines, which was complicated for someone to review, and I was asked to do smaller pieces. 

> Finally, I personally think it's best to have an old component, then a new one, then to ask ppl to move from the old to the new, then to deprecate the old one. It let people the time to test the new one, make suggestions, explain what is good or bad in the new one, what they like in the old one they cannot do in the new one, etc.. This way, the new one can replace properly the old one. I feel like imposing a lot of changes without letting people to get used to the new component is brutal, and can create a lot of frustration. This third point is a personal point of view and I can understand it is not shared.
> 

I agree, and I'm not sure here what would be the best approach because each has it benefits and disadvantages. I guess it would be nice if big changes like these would be able to make it out like a "beta" for a component without having to have a completely different name. There will always be those that don't read release notes and will end up with an impact further down the line.

You could also re-write the component without providing new functionality, just fixes and better code structure & compatibility with climate 1.0 and then add the changes after that. 

> I'm not familiar with HA hierarchy. Who can take a decision on this topic ? Because if I need to remake this PR to replace generic_thermostat it's not complicated.
> 

While I'm not going to add 'tags' here, I do think that the direction would be best up to Owner based also on the consensus from the community. 

> Have a nice day

I do think that this PR can continue in its current format to review code style, functionality, tests & such and if possible (I'll try to help with contribution) add a config flow. The only thing to always keep in mind is not to make it impossible to replace `generic_thermostat` with this.

kevinkk525 commented:
I'm really looking forward to this virtual thermostat implementation. The current generic_thermostat is a really broken pain to use. Just woke up to a cold house because homeassistant rebooted and generic_thermostat lost its away=False temperature..

Personally I think it would be better to introduce it as a new component and mark the generic_thermostat as deprecated and remove it in a future release.
That gives everyone the time to move to the new component.
Alternatively it could replace generic_thermostat but in that case it would need to be able to replace it with the exact same functionality and configuration options and that might be unneccessarily complicated.

Thank you for creating that component! I'm looking forward to using it.

DB-CL commented:
Hi @ciotlosm, 

I'm glad we understand each other in this very polite and respectful way. Thanks for having this conversation.

I won't respond to everything as I understand your opinion, I explained mine, and I have no further comments.

About this idea :
> You could also re-write the component without providing new functionality, just fixes and better code structure & compatibility with climate 1.0 and then add the changes after that.

I won't do it. I started to dig into the code to add functionalities (support of multiple presets) that I needed. I ended up rewriting the component but that was not my first motivation, and that would be to difficult for me to let go that presets support.

I'll let people review the PR as a new component, knowing that I'm ready to make change to the configuration file, or to the way this code arrive into home assistant (beta component, replacement of generic thermostat, new component ..)

On my side, I'll compare generic_thermostat and this PR to check if there is something obvious to make the components more "compliant".

@kevinkk525 thank you for your addition and your support. That strengthens me in my view that creating a new component is a good way to go.

antonverburg commented:
Hi all, I am quite interested in this virtual thermostat. What about a PID regulator, would it be a bennefit if we can add that?

Shulyaka commented:
Why don't we call it a 'template climate', by the way?

b4dpxl commented:
> Why don't we call it a 'template climate', by the way?

Because there's no template involved?

Shulyaka commented:
A PID regulator would be very interesting.
But if it is too complicated to calibrate, I would add a max_cycle_duration and min_idle_duration. This is for heaters that can heat so quickly that we need some time for the heat to reach the sensor.

b4dpxl commented:
> Hi all, I am quite interested in this virtual thermostat. What about a PID regulator, would it be a bennefit if we can add that?

That would need to be _very_ configurable. AFAIK most heating/cooling systems that would be readily controlled by something like this are binary, either on or off. Trying to do something "fake" PID e.g. by PWM would probably fry most domestic heating/cooling devices. I know my boiler wouldn't like it.

antonverburg commented:
Slow PWM is also PWM. I think about having an output value between 0 and 100 or something like that. This number can be intepreted by something outside the thermostat. If there is a servo controlled valve that can be set and controlled, one could set the amount of opening. But if you want to heat the boiler periodically, you could specify the amount of time the boiler should be heating every (for example) one or two hours. In the case here in my home I would like to set the floor heating with such a 'slow pwm' on and off, but I also have a vacuum-tube collector, that has a PWM controlled pump. That should be typically regulated by a direct value. Just some ideas, I am currently pulling this virtual_thermostat on board of my system to see what I can contribute.

b4dpxl commented:
We've already got that with the tolerance values and min_cycle_duration, in some fashion. My heating is binary; although I do have individual control over which radiators are on, I don't have any control over how much they are on or how hot the boiler is. The tolerances control when it comes on and goes off though. I'm concerned that if we make this component overly complex, it's going to introduce more issues of the sort this fork was intended to fix, and make it complex to maintain. Maybe a separate component would be better for a pwm-based thermostat? `simple virtual thermostat` and `pwm virtual thermostat` or something?

antonverburg commented:
Here is somebody who already wrote a PID regulator for thermostat: https://github.com/fabiannydegger/custom_components. I'll compare all and let know if we might be able to combine something.

antonverburg commented:
I've been reading some code. Just some ideas for discussion:
- Why do we have hvac_mode? We want to regulate a certain temperature. If temperature exceeds setpoint, a cooling device should be switched on, if it is lower a heating device has to be switched on. I think the thermostat doesn't even have to know about what heating or cooling devices: a user can make an automation that will start a cooling or heating device depending on the output of the thermostat,
- I think in generic it's a good idea to store the status of the thermostat, so that it will be restored on a reboot. You don't want the thermostat to loose its's setpoint when system is restarted. In my current system I am working on a google agenda, containing an item per room in our house. For each moment I can program when what room has to be heated, using re-occuring items saves me from programming all day separate. Currently I run periodically a script in the background that will start a device or thermostat a certain (configurable) period before the targeted room has to be on temperature, so that heating starts in time to be heated on time. What if we integrate such a function in the thermostat? I think that's someting more interesting for the thermostat to know than 'what device do I control'. But maybe we have to split, and use the thermostat only and alone for temperature regulation, and do other smart stuff outside of it. 
- More or less the same applies to the 'keep-alive' functon. Is this something that belongs in a thermostat? I would think this is something for the climate-decice driver like thing, a thermostat does not want to 'know' about that...

b4dpxl commented:
> Why do we have hvac_mode? 
Because when the HA team implemented “Climate 1.0”, this was the approach they wanted. 🤷‍♂ 

> I think in generic it's a good idea to store the status of the thermostat, so that it will be restored on a reboot. You don't want the thermostat to loose its's setpoint when system is restarted. 
This should be a given for any component

> More or less the same applies to the 'keep-alive' functon. Is this something that belongs in a thermostat? I would think this is something for the climate-decice driver like thing, a thermostat does not want to 'know' about that...
What “climate-device”? There is no such thing with generic_ or virtual_thermostat. That’s largely the point of these components, to allow you to define a climate device using “dumb” sensors and switches. In my case, it drives a Shelly 1 switch, which has zero controls other than “turn on” and “turn off” upon request. It’s purely a wifi-enabled switch, designed primarily for lights.

antonverburg commented:
To my feelings too much functions are integrated in the thermostats. Wouldn't  it be better if we create an integration called something like 'slow_pwm', that will handle the conversion of 0..100percent output from a PID regulator and convert that to a slow pwm regulated output with configurations like frequency, min_on and min_off time? I have the same feeling for the keep_alive switch. That is something that an integration for the particular device should cover, not a thermostat. The thermostat should only have a setpoint, actual value, output value and the PID settings (Kp, Ki, Kd). Optionally we could make it more fun by adding auto-tuning functions, but basically that's all we need. To who do I need to talk to discuss these (architectural) decissions?

antonverburg commented:
BTW: having an  hvac_mode might be a good idea, as one can expect that the regulator parameters should be different for heating or cooling, as heating- and cooling devices can be different in power output.

DB-CL commented:
Hi, 

I'm glad to see some activity around this PR, I'm still waiting for a review ;)

I will try to be concise and quote only the main ideas, that does not mean I didn't read all, thank you for your comments.

> What about a PID regulator, would it be a bennefit if we can add that?

I've been taught that PID with simple binary controls create some instability and is difficult to use, but I've never experimented it and that would be very interesting to do ! Still, I would like this PR to be accepted before adding new features, to not disturb the review process.

> Why do we have hvac_mode? 

You describe an "auto" mode. Which is also an HVAC_MODE in Climate 1.0 (called heat_cool if I'm not mistaken). We could easily add this feature later on, as a support for a new HVAC_MODE in this thermostat to let people choose the behavior they want.

> To my feelings too much functions are integrated in the thermostats.

I disagree on this point. The way you describe (a simple component and add the smartness elsewhere) does not seems to fit well with HASS and its integration system. The components are standalone and does not depend each other, and I personally think this is a strength. That does not mean that what you say is wrong, we could benefit more re-usability between components. But today, it's done that way, and it also has its advantages ( like we don't have to convince the entire world to make a modification ). Each component is a mini-project, and the more isolated they are, the easier it is to maintain them. Maybe the way to go is to do a better job on the component itself, like spliting the files into more objects, and have a better code organisation into the component.

> To who do I need to talk to discuss these (architectural) decissions?

I don't know, I'm still having a hard time understanding Hass internal organisation.

antonverburg commented:
> > What about a PID regulator, would it be a bennefit if we can add that?
> 
> I've been taught that PID with simple binary controls create some instability and is difficult to use, but I've never experimented it and that would be very interesting to do ! Still, I would like this PR to be accepted before adding new features, to not disturb the review process.

Clear.
> 
> > Why do we have hvac_mode?
> 
> You describe an "auto" mode. Which is also an HVAC_MODE in Climate 1.0 (called heat_cool if I'm not mistaken). We could easily add this feature later on, as a support for a new HVAC_MODE in this thermostat to let people choose the behavior they want.
Also clear now.
> 
> > To my feelings too much functions are integrated in the thermostats.
> 
> I disagree on this point. The way you describe (a simple component and add the smartness elsewhere) does not seems to fit well with HASS and its integration system. The components are standalone and does not depend each other, and I personally think this is a strength. That does not mean that what you say is wrong, we could benefit more re-usability between components. But today, it's done that way, and it also has its advantages ( like we don't have to convince the entire world to make a modification ). Each component is a mini-project, and the more isolated they are, the easier it is to maintain them. Maybe the way to go is to do a better job on the component itself, like spliting the files into more objects, and have a better code organisation into the component.

OK. I did not know about this philosophy. I have my own project on my hands here, and I was planning to split it in 3 components:
- A slow_pwm component to drive a  pin with a certain min_on time and min_of time for a pwm_cycle time. This could also be used separately, for example to dive floor heating groups open-loop if they do not have a temperature sensor, or to heat something modulated, etc.
- A PID thermostat for the rooms that have a temperature sensor. 
- A scheduler that is able to predict the amount of time required to heat a room to a certain temerature. I have something like that currently in a python script that reads out the google-calendar integration and programs the heaters based on that data. The scheduler should be able to drive both the thermostat or an open-loop slow_pwm device. 

But if dependency between integrations is not allowed it does not make sense to separate this; in that case I have to bring it all together in one component. I have to re-think about that... I preferred to create generic and relative abstract devices that can be easily re-used in different projects.

antonverburg commented:
H, how should I read this documentation about dependencies if integrations do not depend on each other?  https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

Shulyaka commented:
I believe it is not true that components cannot depend on each other.

antonverburg commented:
Another thing to keep in account when introducing the new thermostat: https://github.com/home-assistant/home-assistant/pull/27308. Bug report https://github.com/home-assistant/home-assistant/issues/22535.

antonverburg commented:
I hope next week I will have some time to start working on the slow_pwm pin. I will start with that. Next item is to create a PID regulated thermostat. I will start with a simple PID regulator, no autotune. If that proves to work reliable and stable, we can add fancy features. For now I will do my scheduling with a scipt, if I have time I will add that also as an integration.

pvizeli commented:
We speak about different kind of integration. Level1 they implement just the interface of an API. Level2 with no real hardware or API behind and are include own logic. Generic and this are Level2 integration. However, it's limited how many logic integration are around for an component because we want avoid to much logic stuff inside a integration. This integration can only be merged if they replace the old generic thermostat. So my change request is that you replace the generic thermostat.

Shulyaka commented:
> We speak about different kind of integration. Level1 they implement just the interface of an API. Level2 with no real hardware or API behind and are include own logic. Generic and this are Level2 integration. However, it's limited how many logic integration are around for an component because we want avoid to much logic stuff inside a integration. This integration can only be merged if they replace the old generic thermostat. So my change request is that you replace the generic thermostat.

Can't it be replaced later as described earlier in this thread? I.e. first introduce this component, allow users time to migrate, then remove deprecated one on a later iteration?

antonverburg commented:
> We speak about different kind of integration. Level1 they implement just the interface of an API. Level2 with no real hardware or API behind and are include own logic. Generic and this are Level2 integration. However, it's limited how many logic integration are around for an component because we want avoid to much logic stuff inside a integration. This integration can only be merged if they replace the old generic thermostat. So my change request is, that you replace the generic thermostat.

Hi Pvizeli, 
* what about my proposal to have 3 different integrations that each implement a part of what we want to use in future? 
* They are all Level 2 integrations as far as I see, should I pack them all in one integration to limit the amount? 
* Is it beter to  convert this PR to the code of those 3 components (including a PID thermostat), and review that in one time & replace the generic thermstat with that new components, or do you prefer the softer roadpath of having generic thermostat and vitual thermostat side-by-side, and later depreciate the generic thermostat replacing it with the virtual one, and after that replacing the vitual thermostat with PID thermostat by depreciating the virtual thermostat?

pvizeli commented:
Well, if you want this on core, replace the generic thermostat and we make an breaking change. Otherwise close this PR and use it as custom component.

pvizeli reviewed (changes requested):
This integration can only be merged if they replace the old generic thermostat. Rename this to `generic_thermostat` and make a breaking change.

rdehuyss commented:
@DB-CL I really hope you continue your hard labor and we can replace the generic thermostat with your implementation. Thanks again for all your effort and know that there are a lot of people supporting you!

Misiu commented:
@DB-CL this already looks awesome. Thank you for the massive amount of work. I also hope that your implementation can replace the `generic thermostat`.
@b4dpxl had some good suggestions, I see that you implemented some.
I'm really looking forward to seeing this! Keep up the good work!

DB-CL commented:
Hi, 

Thanks all for your support and sorry for not beeing around these days, I had a lot to do.
I guess I should ignore all the comments marked as spam.
I'll work on the breaking changes.

akasma74 commented:
It would be great to have a more functional and reliable thermostat integration so thanks for your hard work @DB-CL !

DB-CL committed:
Replacement of generic thermostat

DB-CL committed:
Cleaning

DB-CL committed:
Lint

DB-CL committed:
More lint

DB-CL commented:
Hi !

I've done the adpatations to make a breaking change and totally replace the generic thermostat.

I'm wondering : should I start a new PR since the description of this one is now a bit off, or should we stay with this PR ?

I'll adapt the documentation later on, depending on what we decide for this PR

MartinHjelmare commented:
Please submit a new PR to get a clean thread without all the comments. Review hasn't started so it's fine. Please link to this PR from the new PR.

springstan commented:
@DB-CL any updates on the new PR? :)

nagyrobi commented:
+1 for this! Please let this in!!!

DB-CL commented:
Hi, 

Sorry all for the delay, I have a newborn at home and this one is eating my free time

Many thanks for your support, I'm working on the PR but I still have the documentation to change.

See you soon on a new PR.

Misiu commented:
> Sorry all for the delay, I have a newborn at home and this one is eating my free time

@DB-CL take your time, the family is the most important thing and time with a newborn is priceless.
Congratulations! 🎉

nagyrobi commented:
Congratulations! I am with you brother, I felt the same a couple years ago!

springstan commented:
Closing this PR due to inactivity and to clean up old PRs.
@DB-CL As noted in https://github.com/home-assistant/home-assistant/pull/27833#issuecomment-573360559 please open a new PR to start the reviewing process :)
